test/test_utils/ec2.py

```
@@ -452,6 +458,73 @@ def get_ec2_instance_tags(instance_id, region=DEFAULT_REGION, ec2_client=None):
+    :param required_log_ending: str, The string that is desired to be present at the end of the logs
+    :param loop_time: int, seconds for which we would wait for the tests to execute on ec2 instance
+    :param loop_sleep_time: int, interval at which the logs would be uploaded to the s3 bucket
+    :param log_location_within_ec2: Location within ec2 instance where the logs are being witten.
+    """
+    connection.run(execution_command, hide=True, timeout=connection_timeout, asynchronous=True)
+    start_time = int(time.time())
+    loop_count = 0
+    local_filename = s3_location.split("//")[-1].replace("/", "-")
+    last_line_of_log = ""
+    line_count_list = []
+    while (int(time.time()) - start_time <= loop_time) or (not last_line_of_log.endswith(required_log_ending)):
+        time.sleep(loop_sleep_time)
+        loop_count += 1
+        connection.run(f"aws s3 cp {log_location_within_ec2} {s3_location}")
+        last_line_of_log = get_last_line_of_s3_file(s3_location, local_filename)
+        number_of_lines_in_log_file = int(run("wc -l conftest.py", hide=True).stdout.strip().split()[0])
+        line_count_list.append(number_of_lines_in_log_file)
+        number_of_previous_line_counts_to_check = 3
+        if len(line_count_list) >= number_of_previous_line_counts_to_check:
+            if all(
+                line_count == line_count_list[-1]
+                for line_count in line_count_list[:-number_of_previous_line_counts_to_check]
+            ):
+                # If last 3 runs lead to sam line number then it demonstrates no progress and 
+                # hence we stop.
+                LOGGER.error("No Progress Reported!!")
+                break
+        last_line_of_log = get_last_line_of_s3_file(s3_location, local_filename)
+        LOGGER.info(f"Uploaded file to {s3_location} for {loop_count} number of times")
+    assert last_line_of_log.endswith(required_log_ending), f"Test failed!! Check {s3_location}"
```

It's arguable but I think you should throw an exception here. E.g. `ValueError("Last row is not as expected")` or something. In some sense it's a runtime error. It's not something you test. 
But even if you keep assert, could you add some explanations to the error message. E.g. "Test failed because the last row is not as expected: 
last row in the log file - {};
expected - {}.
Full log - {s3_location}"

---

Done.